### ISSUE TITLE
Loader Redesign Idea

### ISSUE THREAD

**Issue by spficklin** [tripal/tripal#1595]
# Core Tripal 4 Development Task

## Branch Name: tv4g4-issue1595-data_importing_redesign

In the process of discussing the Pub (PubMed, Agricola, Bibtex) and Tree (newick, taxonomy) importers, we (@dsenalik , @risharde) recognized some challenges.   Some history first.  Issue #1461 provided a design for the pub importer.  We have multiple types of importers that all provided the same information about publications but from different sources with different API/file formats.  The idea was to write parsers for each source seperately that return a single associative array or object that represents the class of data (i.e., publication) and then it gets inserted into the database.  Doug tried to implement this style of importer with the PR #1591 for the tree importer.  In this design, each file type is a plugin (e.g. Tree plugin, Pub plugin).  However, what we determined is that using this model we will be duplicating the same type of plugin code repeatedly, whereas if we created a common plugin format we could reduce this.  

So we came up with the following design: 

### TripalParser Plugin
The TripalParser plugin is generic, it's meant for parsing of any File or API.  Because some parsers need information from the user it will also provide a few form functions that the TripalImporter instance can use.   Here is a list of classes and functions:

- `TripalParser` (Annotation class)
- `TripalParserInterface`
  - `::form()`  # Used by the TripalImporter (if needed) to provide necessary inputs for parsing
  - `::formState()`
  - `::formValidate()`
  - `::parse()`  # Performs the parsing
- `TripalParserManager`
- `TripalParserBase`
  - `::__construct()`
- 

Examples of instances of plugins could include the following. 
- Publication parsers:
  - `PubMedParser extends TripalParserBase`
  - `BibTexParser extends TripalParserBase`
- Tree parsers:
  - `NCBITaxonomyParser extends TripalParserBase`
  - `NewickParser extends TripalParserBase`
- Whole genome feature parsers :
   - `GFFParser extends TripalParserBase`
   - `GTFParser extends TripalParserBase`
- Ontology parsers:
  - `OBOParser extends TripalParserBase`
  - `OWLParser extends TripalParserBase`

These all need to return an associative array/object that is used to define the data class type. 

### TripalDataStorage Plugin
This plugin provides an interface for classes of data (e.g. publications, organisms, trees, etc.).  They talk directly to the database, and unlike the existing `TripalStorage` class, they do not expect integration with an entity's fields.   The plugin provides generic CRUD and validation routines:

- `TripalDataStorage` (annotation class)
- `TripalDataStorageInterface`
  - `::validate()`
  - `::insert()`
  - `::delete()`
  - `::update()`
  - `::find()`
  - `::autocomplete()`  # used to provide an autocomplete for the data class.
  - `::selectOptions()`  # used to populate select drop-downs with values (e.g. organism)
- `TripalDataStorageManager`
- `TripalDataStorageBase`

At the Tripal-level we create instances of plugins for each class of data. These plugins simply need to provide validation of the data class. They won't allow CRUD operations if the provided values are not correct. This way parsers must always provide an associative array or object in a format that all data stores can work with.  Examples include:

- `TripalPubStorage extends TripalRecordStorageBase` (PubMed, BibTex, Agricola)
- `TripalTreeStorage extends TripalRecordStorageBase` (newick, taxonomy)
- `TripalOrganismStorage extends TripalRecordStorageBase` (NCBI taxonomy, yml)
- `TripalGenomeStorage extends TripalRecordStorageBase` (GTF, GFF2, GFF3)

At the data store level the plugin will extend the Tripal version  (with the implemented validation) and provide the the insert, delete, update and find routines. Examples include:
- `ChadoPubStorage extends TripalPubStorage`
- `ChadoTreeStorage extends TripalTreeStorage`
- `ChadoOrganismStorage extends TripalOrganismStorage`
- `Neo4jTreeStorage extends TripalTreeStorage`
- `NoSQLTreeStorage extends TripalTreeStorage`

### How to use these in a TripalImporter
To use these classes in a Tripal importer the developer would need to do the following:
1. Instatiate an instance of the appropriate TripalParser plugin to get the form elements and provide those to the user if there are any.
2. After the user submits the form a job is sumbitted.
3. When the job runs, the importer will instantiate the parser again and pass in the information from the user to parse the file or remote API.  It will then generate the appropriate associate array or object.   
4. After the array/object is created it will then create an instance of the proper `TripalDataStorage` plugin to first validate that the array/object is correct.
5. Then, it will call the `::insert()` function for the data type.  The insert function should be optimized for storing the data object in the database.  

### Command-line import
1. A Drush command would need to be created for importing the data type (e.g. organism, publication)
2. The Drush command will need to support arguments needed by the parser class.
3. The user provides the necessary arguments to the Drush command on the command-line.
4. The Drush command will instantiate the proper parser, then parse the input information.
5. The Drush command will then instantiate the proper `TriapalDataStorage` class to insert the data.
6. Simliar drush commands could be made for updating, deleting.

### Advantages
1.  We can write parsers for any file, that can be used by any importer class. The parser plugin ensures that a parser is easily used by `TripalImporter`.
7. The `TripalDataStorage` plugin allows us to validate and provide CRUD operations for classes of data. There is no limit to how many of these we can create.  The validation can happen at the Tripal-level to ensure that data class types are all of consistent format before being stored by the database level version of the plugin.
8. The `TripalDataStorage` plugin will also allow us to provide API functions for data class types that are no longer flat functions. We should be able to deprecate in a future release of Tripal those stand-alone function files.  
9. The `TripalDataStorage` plugin, because it's not tied to the importer, or the file parser, we can provide Drush command-line functions that alllow users to interact with the database programmatically. This was a wish list years ago for Tripal v3.

**Comment by laceysanderson** [comment]
Whoa, this has huge implications for existing importers in Tripal core and those under development in my group...

Specifically, @carolyncaron is working on a genotypes loader which will support 3 file formats plus it will have both an importer implementation and a drush command. 

She is actually already working on a parser plugin called genotypesloader and has implemented the [base class, interface and plugin manager](https://github.com/TripalCultivate/TripalCultivate-Genetics/tree/4.x/trpcultivate_genotypes/src/GenotypesLoader) and [started a VCF implementation](https://github.com/TripalCultivate/TripalCultivate-Genetics/blob/4.x/trpcultivate_genotypes/src/Plugin/GenotypesLoader/VCFGenotypesLoader.php).

@reynoldtan is working on phenotypes importers. He'll also have two formats and slightly different data storage. Right now he's focusing on the multi-step for though.

**Comment by spficklin** [comment]
It also has big implications for the GFF and OBO loaders... if we decide to rewrite them.  My thinking is we leave existing loaders as is.... New loades can use this new infrastrucuture, especially if there are multiple file types for the same class of data.    We only need to migrate existing loader to use this format if somoene has the time and wants to tackle it.

**Comment by laceysanderson** [comment]
Ok, now feedback on design:

tl;dr 
- I am completely in favour of the TripalParser class. 
- The TripalImporter must not require any additional plugins so as to not make data importers too complicated for groups with less funding/knowledge.
- I think we should hold off until a much later version of Tripal to flesh out the design of the TripalDataStorage.

## TripalParser Plugin

I completely agree that having a parser plugin would save a lot of duplication. As an extension to my above comment I would add the following use cases:

- Genotypic Data:
  - `VariantCallFormatParser`
  - `GenotypeMatrixParser`
  - `GenotypeFlatFileParser`
- Phenotypic Data:
  - `CropFieldPhenotypeParser`
  - `BiochemPhenotypeParser`
  - `GreenhousePhenotypeParser`

### Add set/get methods for common options

It would be good for the base class to have getter/setters for common metadata. Even though a given parser may not use all getter/setters provided having them could save a lot of duplication and ensure consistent validation. See Carolyns [setInputFilePath](https://github.com/TripalCultivate/TripalCultivate-Genetics/blob/4.x/trpcultivate_genotypes/src/GenotypesLoader/GenotypesLoaderPluginBase.php#L330-L353). One thing me + Carolyn discussed was using the file path specifically here rather than a fid to better support drush commands. I think we would also want an setApiUri -again we would provide both and then the importer would use the one it needs and not worry about the other.

This approach allows for more validating or metadata to be taken care of by the base where just passing everything into the parser function requires all the validation there and makes good automated testing harder.

### Add parser datatype to ensure compatibility/discoverability

One thing we would really need to keep in mind is ensuring the plugin has a way to indicate a grouping for parsers so we can link compatible parsers with specific importers/drush commands. Perhaps a `parser_data_type` entry in the annotation. This would allow the plugin manager to find the right parsers without hard-coding the appropriate options and allows other modules to add parsers for a given importer. This code does something similar in  Carolyn plugin manager:

https://github.com/TripalCultivate/TripalCultivate-Genetics/blob/4.x/trpcultivate_genotypes/src/GenotypesLoader/GenotypesLoaderPluginManager.php#L55-L63
```php
    // Chooses plugin implementation based on parameters (e.g. VCF implementation for VCF file format)
    // Collect all the plugin IDs that match our input file type
    $supported_pluginIDs = [];
    $definitions = $this->getDefinitions();
    foreach($definitions as $pluginId => $details) {
      if ($details['input_file_type'] == ($options['input_file_type'])) {
        $supported_pluginIDs[] = $pluginId;
      };
    }
```

### Define required input rather than provide form elements

I would suggest having a public variable where the parser defines the metadata it needs for input rather then the form/formValidate/formSubmit functions (or at least in addition to them.

For example, say our example parser needs an organism to be set for passing to an API:
```php
public array $metadataDefinition = [
  'organism' => [
    'data_type' => 'string',
    'element_type' => 'textfield',
    'label' => 'Species',
    'help_text' => 'Indicate the species you want to restrict publication to by including the full scientific name.'
  ],
];
```

This way the base class could generate the form elements based on the data definition just as I do [here in the this Tripal3 module](https://github.com/UofS-Pulse-Binfo/chado_custom_search/blob/master/chado_custom_search/ChadoCustomSearch.inc#L199-L239). This way the drush command could similarly build prompts based on this information and it would remain in sync. Each parser could override the form method if needed, call the parent to build the base and then only make specialized changes needed.

Back to my example above, the form method in the base would generate a basic textfield, then the specific parser would overide form(), call the parent and set the autocomplete path.

## TripalDataStorage Plugin

I see where you are going with this and agree with it in theory.

In practice I'm worried about a couple of things which make me think we should not approach this now but maybe wait a year or two ;-)

### Concerns

#### Having two plugin devoted to data storage CRUD that cannot be used interchangeably is going to cause a lot of confusion.

I agree that you cannot use the current TripalStorage as it is entity/field specific but I do not think just writing another one is a good idea 😬 Especially a we may want something like this again for the search API.

Instead I think we may be better served to create a common base plugin that can serve all three of these needs or at least can keep common code and then be extended to serve each need independently.

Of course, now is not a good time to take this on because we do not have the developer time or knowledge to design this properly at this time. Also it's a big undertaking and I think we need to be less idealistic in order to get a working Tripal version out ASAP.

#### I worry about putting the autocomplete/select in TripalDataStorage

It feels like this would be a heavy weight way to provide that functionality. Autocomplete/select would be better as independent services in my opinion in order to keep them lightweight and make it as easy as possible to pull them in through dependency injection into any form or page.

**Comment by spficklin** [comment]
**Responses to ideas on TripalParser**
> Add set/get methods for common options

I like this suggestion.  

> Add parser datatype to ensure compatibility/discoverability

I'm not sure why we would need this.  If the Importer class already knows what file formats it supports then why would we need to group them using a class member variable?  Adding a group variable doesn't seem quite right to me... not sure why.... Shouldn't we be doing something more along the lines of plugins rather than setting a variable for grouping?

> Define required input rather than provide form elements

I see the value in this, but it means we have to write infrastructure in Tripal to support this and a developer has to learn this too.  I'm not opposed to it, but I think we need to think on this a bit more to make sure we're not creating more work for extension developers.  Isn't the `validate()` and `formValidate()` sufficient? 

**Response to concerns on TripalDataStorage**

> I agree that you cannot use the current TripalStorage as it is entity/field specific but I do not think just writing another one is a good idea 😬 Especially a we may want something like this again for the search API. Instead I think we may be better served to create a common base plugin that can serve all three of these needs or at least can keep common code and then be extended to serve each need independently.... Of course, now is not a good time to take this on because we do not have the developer time or knowledge to design this properly at this time. Also it's a big undertaking and I think we need to be less idealistic in order to get a working Tripal version out ASAP.

We're already writing the code to do the SQL statements inside of the importers for Chado..  The idea here is just to move that out of the importers so we can build other backends (e.g. Neo4j, NoSQL) to support loading without having to always update the loaders.   I just took it one step further and added update/edit functions and allow it to provide some validation and suck up some stand-alone API functions for similar data types.    Basically, the idea is to move the loading of the data out of the importers so we don't need a  TreeImporte rin the `tripal_chado` module and another one in the `tripal_noSQL` module, etc.  

I'm not sure I understand what 3 cases we should support with the ChadoStorage class.  I don't think we should expand the ChadoStorage class to generically support loading these type of data because we have to do some very specific things to make bulk inserts work quick enough.  Each  loader has to do very different things to get complex data structures into the backed in a fast way. I don't know how we could write code to support that in the ChadoStorage class.

**Comment by laceysanderson** [comment]
> > Add parser datatype to ensure compatibility/discoverability
> 
> I'm not sure why we would need this. If the Importer class already knows what file formats it supports then why would we need to group them using a class member variable? Adding a group variable doesn't seem quite right to me... not sure why.... Shouldn't we be doing something more along the lines of plugins rather than setting a variable for grouping?

I've very confused by this comment...
It sounds like the goal is to write the publication importer for example, to know which parsers it supports... but then how do we discover parses written by extension modules? 

So if in core we implement the PubMedParser and BibTexParser parsers for the publication importer... then we write the publication importer to use them. What if an extension module implements a new parser type (e.g. agricola)? How would the Tripal publication importer know that should be added to the publication importer options?

Writing an entirely new plugin type to handle supporting extension module parses kind of defeats the purpose of having a shared TripalParser plugin 🤔 

Also, the datatype would be set in the annotation of the tripalparser implementation. I meant a variable in the annotation class in case that wasn't clear. So it is a typical plugin thing to do 🤷‍♀️

**Comment by spficklin** [comment]
Sounds like we need an in person discussion to sort this out.  Maybe at an upcoming Monday morning dev session?

**Comment by spficklin** [comment]
Summary of discussion on 8/14/2023 in the Tripal Dev file

- For publications
  -  Create a base class for publication importers
  -  Each importer (bibtex, PubMed, etc) then then we extend the base class -- we get separate loaders for each one.
- We will not implement any of the class proposals in this issue as they will create complicated abstraction problems, and they will create more work for us in the long haul.

**Comment by laceysanderson** [comment]
Slight expansion of the following point for anyone who finds this later

> We will not implement any of the class proposals in this issue as they will create complicated abstraction problems, and they will create more work for us in the long haul.

Some of these complicated abstraction problems are:

GFF3 importer needs to keep track of previously imported parent records (e.g. gene) when importing child records (e.g. mRNA, exon). It has done this by caching these records in a file-based cache for use later. It also uses bulk inserts which speed things up greatly. 

Also for the parser we would need to keep large files in mind, simply returning the whole file in an array form would cause large memory usage and could cause crashes.

VCF importer (extension) caches genotypic data in a file and then copies it in in even larger bulk operations for speed considerations because of the size of these files.
